### PR TITLE
Efficient renames (fixes #1217)

### LINKED ISSUE #1217: More efficient renaming
**normal scenario**
- folder A (e.g. 100 files / 50GB) on computer X is synced to computer Y
- on computer X folder A is renamed to folder B and synced
- Syncthing on computer Y
  - copies existing files (reusing blocks) from folder A to folder B - total size 100GB
  - deletes folder A - total size 50GB
:+1: 

**out of space scenario** (e.g. 6GB free space)
- folder A (e.g. 100 files / 50GB) on computer X is synced to computer Y
- on computer X folder A is renamed to folder B and synced
- Syncthing on computer Y
  - copies existing files (reusing blocks) from folder A to folder B
  - stops copying as disk is full (e.g. total size 56GB)
  - deletes all 100 files folder A
- admin frees up space
- Syncthing needs to transfer 44GB

expected
do not delete old files if running out of space until all changes are made and all blocks are reused

### PR DESCRIPTION
**Pull request by AudriusButkevicius**
Just did a basic test renaming folders, seemed to work.

A possible change would be not to take this shortcut if versioning is enabled, and let the file to be assembled from blocks, as the versioned file has to be copied to the archive anyway.

### CODE REVIEW

**Comment by AudriusButkevicius** [comment]
http://build.syncthing.net/job/syncthing-integration/216/

File: internal/model/puller.go [hunk 1]
```diff
@@ -367,13 +398,12 @@ func (p *Puller) pullerIteration(ignores *ignore.Matcher) int {
...
 
-	for i := range deletions {
-		deletion := deletions[len(deletions)-i-1]
-		if deletion.IsDirectory() {
-			p.deleteDir(deletion)
-		} else {
-			p.deleteFile(deletion)
-		}
+	for _, file := range fileDeletions {
+		p.deleteFile(file)
```
Comment (calmh) [inline internal/model/puller.go]: We could change the signature of deleteFile to only take the name, and then the map only needs to be `map[string]struct{}`

File: internal/model/puller.go [hunk 2]
```diff
@@ -479,6 +509,41 @@ func (p *Puller) deleteFile(file protocol.FileInfo) {
...
+	to := filepath.Join(p.dir, target.Name)
+
+	var err error
+	if p.versioner != nil {
+		err = osutil.Copy(from, to)
+		if err == nil {
+			err = osutil.InWritableDir(p.versioner.Archive, from)
+		}
+	} else {
+		err = osutil.InWritableDir(func(_ string) error {
```
Comment (calmh) [inline internal/model/puller.go]: TryRename already does this as part of withPreparedTarget, no?

File: internal/model/puller.go [hunk 3]
```diff
@@ -479,6 +509,41 @@ func (p *Puller) deleteFile(file protocol.FileInfo) {
...
+			return osutil.TryRename(from, to)
+		}, to)
+	}
+
+	if err != nil {
+		l.Infof("Puller (folder %q, file %q): rename from %q: %v", p.folder, target.Name, source.Name, err)
+		return
+	}
+
+	if !target.IsSymlink() {
```
Comment (calmh) [inline internal/model/puller.go]: This stuff could use a comment about how this results in both the new file and the delete getting correctly recorded in the database, and what the symlink case is about... I.e. that source already is supposed to have the delete bit set etc.

File: internal/model/puller.go (hunk 1 above)
Comment (AudriusButkevicius) [inline internal/model/puller.go]: We get the file from `WithNeed` anyway, so I don't see a lot of benefit doing additional DB lookups? 
Are you worried that adding the value to the map copies the struct?

File: internal/model/puller.go (hunk 3 above)
Comment (AudriusButkevicius) [inline internal/model/puller.go]: `Maybe even make this method work on just the file names` that would mean db lookups inside the function, though I already have both FileInfo's available.

As to the symlink case, I guess I don't really need the special case handling now that I think about it.

**Comment by AudriusButkevicius** [comment]
:warning:  Integration tests failing

File: internal/model/puller.go (hunk 1 above)
Comment (calmh) [inline internal/model/puller.go]: I was thinking that we don't actually need the FileInfo at all to do the delete, but we do (since deleteFile calls updateLocal), so ignore that comment.

**Comment by AudriusButkevicius** [comment]
So this seems to pass now:
http://build.syncthing.net/job/syncthing-integration/220/console

**Comment by calmh** [comment]
I don't think the integration tests actually do any renames. That should be added.

**Comment by AudriusButkevicius** [comment]
Agree, though first step was getting the existing ones passing ;)
I'll add some.

**Comment by calmh** [comment]
@st-jenkins test this please

**Comment by AudriusButkevicius** [comment]
Still failing

**Comment by calmh** [comment]
http://build.syncthing.net/job/syncthing-integration/ws/test/2.out/*view*/

Index out of range

**Comment by AudriusButkevicius** [comment]
Aaaand how is that possible?

File: internal/model/puller.go [hunk 4]
```diff
@@ -341,12 +353,33 @@ func (p *Puller) pullerIteration(ignores *ignore.Matcher) int {
...
 			break
 		}
 		if f, ok := p.model.CurrentGlobalFile(p.folder, fileName); ok {
+			if !f.IsSymlink() {
+				key := string(f.Blocks[0].Hash)
+				for i, candidate := range buckets[key] {
+					if scanner.BlocksEqual(candidate.Blocks, f.Blocks) {
+						// Remove the candidate from the bucket
+						l := len(buckets[key]) - 1
+						buckets[key][i] = buckets[key][l]
```
Comment (Zillode) [inline internal/model/puller.go]: Errors if only one element is in the buckets:
l = 1;
_[0] = _[1];

Comment (AudriusButkevicius) [inline internal/model/puller.go]: No, l in that case would be 0, then we'd do:
_[0] = _[0]
_ = _[:0]
making _ an empty slice.

The error is at line 339, which is file.Blocks[0], somehow the file which comes out of the database has no blocks.

File: internal/model/puller.go [hunk 5]
```diff
@@ -327,7 +329,17 @@ func (p *Puller) pullerIteration(ignores *ignore.Matcher) int {
...
 			// A deleted file, directory or symlink
-			deletions = append(deletions, file)
+			if file.IsDirectory() {
+				dirDeletions = append(dirDeletions, file)
+			} else {
+				fileDeletions[file.Name] = file
+				df, ok := p.model.CurrentFolderFile(p.folder, file.Name)
+				if ok {
+					// Put files into buckets per first hash
+					key := string(df.Blocks[0].Hash)
```
Comment (calmh) [inline internal/model/puller.go]: Out of range here, so df has no blocks apparently. Maybe print it at figure out what it is. Already deleted, but then why should we delete it again?

**Comment by AudriusButkevicius** [comment]
Success at last.
http://build.syncthing.net/job/syncthing-integration/lastSuccessfulBuild/

File: test/util.go [hunk 6]
```diff
@@ -116,6 +116,10 @@ func alterFiles(dir string) error {
 			return err
 		}
 
+		if strings.Contains(filepath.Base(path), "test-") {
```
Comment (calmh) [inline test/util.go]: Maybe strings.HasPrefix, just to be really picky here?

File: internal/model/puller.go [hunk 7]
```diff
@@ -341,12 +353,33 @@ func (p *Puller) pullerIteration(ignores *ignore.Matcher) int {
...
+						// Remove the pending deletion (as we perform it by renaming)
+						delete(fileDeletions, candidate.Name)
+
+						p.renameFile(candidate, f)
+
+						p.queue.Done(fileName)
+						goto nextFile
+					}
+				}
+
```
Comment (calmh) [inline internal/model/puller.go]: Spurious space

File: internal/model/puller.go [hunk 8]
```diff
@@ -341,12 +353,33 @@ func (p *Puller) pullerIteration(ignores *ignore.Matcher) int {
...
 	})
 
+nextFile:
 	for {
 		fileName, ok := p.queue.Pop()
 		if !ok {
 			break
 		}
 		if f, ok := p.model.CurrentGlobalFile(p.folder, fileName); ok {
+			if !f.IsSymlink() {
```
Comment (calmh) [inline internal/model/puller.go]: Nest gets deep here. Take a quick exit instead?

```
if f.IsSymlink() {
p.handleFile(...)
continue;
}
```

File: internal/model/puller.go [hunk 9]
```diff
@@ -341,12 +353,33 @@ func (p *Puller) pullerIteration(ignores *ignore.Matcher) int {
...
+						l := len(buckets[key]) - 1
+						buckets[key][i] = buckets[key][l]
+						buckets[key] = buckets[key][:l]
+						// Remove the pending deletion (as we perform it by renaming)
+						delete(fileDeletions, candidate.Name)
+
+						p.renameFile(candidate, f)
+
+						p.queue.Done(fileName)
+						goto nextFile
```
Comment (calmh) [inline internal/model/puller.go]: I think `continue nextFile` sounds nicer and illustrates that we're taking a new round of a loop, not just going someplace random for vacation. :)

File: internal/model/puller.go (hunk 8 above)
Comment (calmh) [inline internal/model/puller.go]: Also, symlink needs `p.queue.Done()` I think?

Comment (AudriusButkevicius) [inline internal/model/puller.go]: yeah, that's a bit hard to follow, but not sure how to make it better.

**Comment by AudriusButkevicius** [comment]
http://build.syncthing.net/job/syncthing-integration/232/console
